app/views/service_worker/service_worker.js

```diff
@@ -11,10 +11,33 @@ self.addEventListener('push', async (event) => {
   const title = data.title
   const options = {
     body: data.body,
-    icon: data.icon
+    icon: data.icon,
+    data
   }
 
   event.waitUntil(
     self.registration.showNotification(title, options)
   )
 })
+
+self.addEventListener('notificationclick', function (event) {
+  event.notification.close()
+  if (!event.notification.data.path) return
+
+  event.waitUntil(
+    clients.matchAll({ type: 'window' }).then((clientList) => {
+      for (let i = 0; i < clientList.length; i++) {
+        const client = clientList[i]
+        const clientPath = (new URL(client.url)).pathname
+
+        if (clientPath === event.notification.data.path && 'focus' in client) {
+          return client.focus()
+        }
+      }
+
+      if (clients.openWindow) {
+        return clients.openWindow(event.notification.data.path)
+      }
+    })
+  )
+})
```

Comment:
If the `pathname` of the client matches the path stored in the notification's data, it focuses on that client (brings the tab/window to the foreground) instead of opening a new one. See this [Reference](https://github.com/rails/rails/pull/50528/files).